src/typing/flow_js.ml

```diff
@@ -4236,6 +4239,18 @@ let rec __flow cx ((l: Type.t), (u: Type.use_t)) trace =
       let new_obj = ObjT (reason, {objtype with flags}) in
       rec_flow_t cx trace (new_obj, t)
 
+    (*********************************************************)
+    (* objects can be marked as read-only using $ReadOnly<T> *)
+    (*********************************************************)
+
+    | (ObjT (reason_o, objtype), UseT(_, ReadOnlyT (reason_op, t))) ->
+      let desc = RReadOnly (desc_of_reason reason_o) in
+      let reason = replace_reason_const desc reason_op in
+
+      let flags = {frozen = true; sealed = Sealed; exact = true;} in
+      let new_obj = ObjT (reason, {objtype with flags}) in
```

Comment:
That's fair. Also as @mroch mentioned, this remains unsound since it creates a new type representing the same object as a previous type that is more lenient (mutable) -- which means the new type's assertions that things can't change could be wrong if the previous one is mutated.

Anyway: I'm going to be out on leave for about a month starting soon, so it probably makes sense to hand this over to @mroch to decide how to move forward. Thanks a ton for starting the discussion!